tokio-udp/src/socket.rs

```diff
@@ -10,6 +10,12 @@ use mio;
 use tokio_reactor::{Handle, PollEvented};
 
 /// An I/O object representing a UDP socket.
+///
+/// **Note**: Well `UdpSocket` is `Sync`, the caller must ensure that there are
```

Comment:
I think you ment: "While `UdpSocket` is `Sync`".

However why not keep the methods require a mutable reference? This way it's enforced in the type system that the socket is not used concurrently across threads.

Reply:
It also does not match the API of other IO types (`TcpStream`, `UnixStream`, `UnixDatagram`, etc.), which can be used with a non-mutable reference.

Reply:
@sopium that is true, most types in the net module only require `&self` for its methods. However if the Tokio runtime can't handle multiple callers, wether it be from different tasks or threads, maybe they shouldn't. I think, and I could be wrong here, that those method only take a reference because is what the standard library does, because from an OS perspective it's ok to make system calls concurrently on the same file descriptor.

But the standard library doesn't have any other fields it needs to update in user space, a `TcpStream` or `TcpListener` is just a file descriptor (at least on Unix). I don't believe this is the case for Tokio.